Make to sure to reregister service worker clients when launching a service worker after a network process crash

youennf opened the pull request:
#### b707db9ba55dbed587a5c6043c17216efcddc612
<pre>
Make to sure to reregister service worker clients when launching a service worker after a network process crash
<a href="https://rdar.apple.com/141159225">rdar://141159225</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284302">https://bugs.webkit.org/show_bug.cgi?id=284302</a>

Reviewed by Ben Nham.

When a network process crashes, web processes only create a connection to the network process when needed.
The consequence is that service worker clients may not all be registered until the process owning these clients creates a connection to the network process.

To fix that issue, when we launch a service worker process, we will make sure that all processes that have pages related to the service worker top origin recreate their network process connection.

Covered by added test and updated test.
Skipping added test on iOS like other tests using showNotification in service worker.

* LayoutTests/http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash-expected.txt: Added.
* LayoutTests/http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash.html: Added.
* LayoutTests/http/tests/workers/service/resources/registerServiceWorkerClient-after-network-process-crash-worker.js: Added.
(async const):
(let.messageClients):
(async await):
* LayoutTests/http/tests/workers/service/shownotification-allowed.html:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::networkProcessDidTerminate):
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldRegisterServiceWorkerClients const):
(WebKit::WebProcessProxy::registerServiceWorkerClients):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::registerServiceWorkerClients):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/287757@main">https://commits.webkit.org/287757@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/852be66501c532e766e9cc49fd153e6afd07733f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80711 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/233 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85238 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31695 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63036 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20829 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83780 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/94 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/73467 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43338 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30 "Build is in progress. Recent messages:") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27627 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30152 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86670 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7938 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71329 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70570 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17583 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14591 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7900 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13421 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7739 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -97,6 +97,7 @@ messages -> NetworkConnectionToWebProcess WantsDispatchMessage {
 
     EstablishSWContextConnection(WebKit::WebPageProxyIdentifier webPageProxyID, WebCore::Site site, std::optional<WebCore::ScriptExecutionContextIdentifier> serviceWorkerPageIdentifier) -> ()
     CloseSWContextConnection()
+        PingPongForServiceWorkers() -> (bool isSuccess)
```
nit: fix indentation

bnham reviewed (commented):
Seems reasonable at first glance. Will let it run through the tests.

youennf committed:
Make to sure to reregister service worker clients when launching a service worker after a network process crash
rdar://141159225
https://bugs.webkit.org/show_bug.cgi?id=284302

Reviewed by Ben Nham.

When a network process crashes, web processes only create a connection to the network process when needed.
The consequence is that service worker clients may not all be registered until the process owning these clients creates a connection to the network process.

To fix that issue, when we launch a service worker process, we will make sure that all processes that have pages related to the service worker top origin recreate their network process connection.

Covered by added test and updated test.
Skipping added test on iOS like other tests using showNotification in service worker.

* LayoutTests/http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash-expected.txt: Added.
* LayoutTests/http/tests/workers/service/registerServiceWorkerClient-after-network-process-crash.html: Added.
* LayoutTests/http/tests/workers/service/resources/registerServiceWorkerClient-after-network-process-crash-worker.js: Added.
(async const):
(let.messageClients):
(async await):
* LayoutTests/http/tests/workers/service/shownotification-allowed.html:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::networkProcessDidTerminate):
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldRegisterServiceWorkerClients const):
(WebKit::WebProcessProxy::registerServiceWorkerClients):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::registerServiceWorkerClients):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/287757@main
